Source/Core/Logger/Logger.swift

```diff
@@ -0,0 +1,302 @@
+            force: force,
+            context: String(describing: S.self),
+            file: file,
+            function: function,
+            line: line
+        )
+    }
+
+    func log<T>(
+        level: LogLevel = .info,
+        _ object: @autoclosure () -> T,
+        tags: [String] = [],
+        force: Bool = false,
+        _ file: String = #file,
+        _ function: String = #function,
+        _ line: Int = #line
+    ) {
+        self.log(
+            level: level,
+            object: object,
+            tags: tags,
+            force: force,
+            context: nil,
+            file: file,
+            function: function,
+            line: line
+        )
+    }
+
+    private func log<T>(
+        level: LogLevel = .info,
+        object: @autoclosure () -> T,
+        tags explicitTags: [String] = [],
+        force: Bool,
+        context: String?,
+        file: String,
+        function: String,
+        line: Int
+    ) {
+        #if !DEBUG
+            guard level != .debug else {
+                return
+            }
+        #endif
+
+        let thread = Thread.isMainThread ? "UI" : "BG"
+        let threadID = pthread_mach_thread_np(pthread_self())
+        let timestamp = self.startTime.elapsed
+        let string = "\(object())"
+
+        self.dispatchGroup.enter()
+        self.queue.async { [weak self] in
+            guard let strongSelf = self else { return }
+            strongSelf.synclog(
+                thread: thread,
+                threadID: threadID,
+                timestamp: timestamp,
+                level: level,
+                string: string,
+                tags: explicitTags,
+                force: force,
+                file: file,
+                function: function,
+                line: line,
+                context: context
+            )
+            strongSelf.dispatchGroup.leave()
+        }
+    }
+
+    private func synclog(
+        thread: String,
+        threadID: mach_port_t,
+        timestamp: Double,
+        level: LogLevel = .info,
+        string: String,
+        tags explicitTags: [String] = [],
+        force: Bool = false,
+        file: String = #file,
+        function: String = #function,
+        line: Int = #line,
+        context: String?
+    ) {
+        guard (self._enabled && self.transports.count > 0) || force else {
+            return
+        }
+
+        let functionName = function.components(separatedBy: "(").first ?? ""
+        let fileName: String = {
+            let name = URL(fileURLWithPath: file)
+                .deletingPathExtension().lastPathComponent
+            let value = name.isEmpty ? "Unknown file" : name
+            return value
+        }()
+
+        var allTags = [functionName, thread, fileName, level.rawValue]
+        allTags.append(contentsOf: explicitTags)
+        if let context = context {
+            allTags.append(context)
+        }
+
+        var shouldOutputToTransports = true
```

Comment:
Could both of these if checks be combined into one? And maybe just get assign and then make it a let instead of a var.

Reply:
ahhh my head started hurting when i tried !!